### PR TITLE
WIP:Featrue41 016/046/074/191/224

### PR DESCRIPTION
**Pull request by varsha-22**
tasks has been covered in this PR:

1.Dynamic routing
2.Generic Table component 
3.Adding sort, and onclick even to table
4.Paginantion in table and edit and delete table record
5.Snackbar and context api

All commented issues has been fixed as well.

### CODE REVIEW

File: src/components/Math/Math.jsx [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      first: 0,
+      second: 0,
+      operator: null,
+      result: null
+    };
+  }
+  handleResultButton = event => {
+    event.preventDefault();
+    const { result, first, second, operator } = this.state;
+    let finalResult = result;
```
Comment (pinku4444) [inline src/components/Math/Math.jsx]: Fix this re-assigning issue.

File: src/components/SelectField/SelectField.jsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+      case "cricket":
+        return <RadioGroups name1={cricket} sport={sports} />;
+        // eslint-disable-next-line no-unreachable
+        break;
+      case "football":
+        return <RadioGroups name2={football} />;
+        // eslint-disable-next-line no-unreachable
+        break;
+      default:
+        return <p></p>;
```
Comment (pinku4444) [inline src/components/SelectField/SelectField.jsx]: send empty string.

File: src/pages/TextFieldDemo/TextFieldDemo.jsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      formIsValid: true,
+      errors: {
+        minNameLength: "name must be at least 3 characters",
+        notEmpty: "name is required"
+      }
+    };
+  }
+
+  textChangeHandler = event => {
+    let nam = event.target.name;
```
Comment (pinku4444) [inline src/pages/TextFieldDemo/TextFieldDemo.jsx]: change variable name "nam" to "name"

File: src/pages/TextFieldDemo/TextFieldDemo.jsx [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+      }
+      if (val !== "" && val < 1) {
+        err = <strong style={strStyle}> Age can not be below the 1</strong>;
+      }
+      if (val !== "" && val > 100) {
+        err = <strong style={strStyle}> Age can not be above the 100</strong>;
+      }
+    }
+
+    this.setState({ errmsg: err });
```
Comment (pinku4444) [inline src/pages/TextFieldDemo/TextFieldDemo.jsx]: Use only one setState function here.

File: src/pages/Trainee/TraineeList.jsx [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+              }
+            ]}
+            handleSort={this.handleSort}
+            handleRemove={this.handleRemove}
+            startEditing={this.startEditing}
+            editIdx={this.state.editIdx}
+            stopEditing={this.stopEditing}
+            handleChange={this.handleChange}
+            columnToSort={this.state.columnToSort}
+            sortDirection={this.state.sortDirection}
```
Comment (pinku4444) [inline src/pages/Trainee/TraineeList.jsx]: Use Destructuring concept.

File: src/routes/Routes.jsx [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+import Trainee from "../pages/Trainee/Trainee";
+import TextFieldDemo from "../pages/TextFieldDemo/TextFieldDemo";
+import InputDemo from "../pages/Inputdemo/inputDemo";
+import Math from "../components/Math/Math";
+import NoMatch from "../pages/NoMatch/NoMatch";
+
+const Routes = () => {
+  return (
+    <Router>
+      <div>
```
Comment (pinku4444) [inline src/routes/Routes.jsx]: remove div from here

**Review by pinku4444 — COMMENTED** [review]
Please fix all issue.

File: src/components/Math/Math.jsx (hunk 1 above)
Comment (varsha-22) [reply src/components/Math/Math.jsx]: fixed

File: src/components/SelectField/SelectField.jsx (hunk 2 above)
Comment (varsha-22) [reply src/components/SelectField/SelectField.jsx]: done sir

File: src/pages/TextFieldDemo/TextFieldDemo.jsx (hunk 4 above)
Comment (varsha-22) [reply src/pages/TextFieldDemo/TextFieldDemo.jsx]: done sir

File: src/pages/TextFieldDemo/TextFieldDemo.jsx (hunk 3 above)
Comment (varsha-22) [reply src/pages/TextFieldDemo/TextFieldDemo.jsx]: done

File: src/routes/Routes.jsx (hunk 6 above)
Comment (varsha-22) [reply src/routes/Routes.jsx]: done sir

File: src/pages/Trainee/TraineeList.jsx (hunk 5 above)
Comment (varsha-22) [reply src/pages/Trainee/TraineeList.jsx]: done sir

**Pull request merged**